scripts/west_commands/sdk.py

```diff
@@ -0,0 +1,368 @@
+
+    def install_sdk(self, args, user_args):
+        if not os.path.exists(args.install_base):
+            self.die(f"{args.install_base} not exists")
+
+        if not os.access(args.install_base, os.W_OK):
+            self.die(f"{args.install_base} not writable ")
+
+        self.inf("Fetching Zephyr SDK list...")
+        releases = self.fetch_all_releases(args.api_url)
+
+        if args.sdk_version == "latest":
+            target_release = releases[0]
+        else:
+            target_release = next(
+                filter(lambda x: x["tag_name"] == args.sdk_version, releases)
+            )
+
+        sdk_dir = self.sdk_basename(target_release)
+
+        target_dir = os.path.join(args.install_base, sdk_dir)
+        if "Windows" == platform.system():
+            setup = os.path.join(target_dir, "setup.cmd")
+            optsep = "/"
+        else:
+            setup = os.path.join(target_dir, "setup.sh")
+            optsep = "-"
+
+        if not os.path.exists(setup):
+            sha256_url = self.sha256_sum_url(target_release)
+            sdk_url = self.minimal_sdk_url(target_release)
+
+            self.inf("Fetching assets list...")
+            resp = requests.get(sha256_url, stream=True)
+            if resp.status_code != 200:
+                raise Exception(f"Failed to download {sha256_url}: {resp.status_code}")
+
+            hashtable = self.sha256_hashtable(resp.content.decode("UTF-8"))
+            sha256 = hashtable[self.minimal_sdk_filename(target_release)]
+
+            self.inf(f"Downloading {sdk_url}.")
+            self.download_and_extract(sdk_url, sha256, args.install_base, sdk_dir)
+
+        else:
+            self.inf(f"{target_dir} exists. Use this.")
+
+        with open(os.path.join(target_dir, "sdk_toolchains")) as f:
+            available_toolchains = [l.strip() for l in f.readlines()]
+
+        cmds = [setup]
+
+        if args.toolchains:
+            if "all" in args.toolchains:
+                cmds.extend([optsep + "t", "all"])
+            else:
+                for tc in args.toolchains:
+                    if tc in available_toolchains:
+                        cmds.extend([optsep + "t", tc])
+                    else:
+                        self.die("toolchain {tc} is not available.\n" +
+                                 "Pelase select list below:\n" +
+                                 "\n".join(available_toolchains))
+
+        if args.hosttools:
+            cmds.extend([optsep + "h"])
+
+        self.dbg("Run: ", cmds)
+
+        result = subprocess.run(cmds)
+
+        if result.returncode != 0:
+            self.die(f"command \"{' '.join(cmds)}\" failed")
+
+        # Associate it so that it can be found.
+        cmds.extend([optsep + "c"])
+
+        self.dbg("Run: ", cmds)
+        ret = subprocess.run(cmds)
+        if ret.returncode != 0:
+            self.die(f"command \"{' '.join(cmds)}\" failed")
+
+    def list_sdk(self, args, user_args):
+        cmake_script = """
```

Comment:
Why embed this in a python script, and not make it an actual CMake file?

Reply:
just put the file in a subfolder and open it

Reply:
I put the embedded source into scripts/west_commands/sdk/findsdk.cmake.